sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -960,17 +961,59 @@ case class GpuRegExpReplace(
   override def second: Expression = searchExpr
   override def third: Expression = replaceExpr
 
-  def this(srcExpr: Expression, searchExpr: Expression, cudfRegexPattern: String,
-      cudfReplacementString: String) = {
-    this(srcExpr, searchExpr, GpuLiteral("", StringType), cudfRegexPattern, cudfReplacementString)
+  def this(srcExpr: Expression, searchExpr: Expression, javaRegexpPattern: String,
+    cudfRegexPattern: String, cudfReplacementString: String) = {
+    
+    this(srcExpr, searchExpr, GpuLiteral("", StringType), javaRegexpPattern, 
+      cudfRegexPattern, cudfReplacementString)
   }
 
   override def doColumnar(
       strExpr: GpuColumnVector,
       searchExpr: GpuScalar,
       replaceExpr: GpuScalar): ColumnVector = {
-    withResource(Scalar.fromString(cudfReplacementString)) { rep =>
-      strExpr.getBase.replaceRegex(cudfRegexPattern, rep)
+
+    def isEmptyRepetition(regex: RegexAST): Boolean = {
+      regex match {
+        case RegexRepetition(_, term) => term match {
+          case SimpleQuantifier('*') | SimpleQuantifier('?') => true
+          case QuantifierFixedLength(0) => true
+          case QuantifierVariableLength(0, _) => true
+          case _ => false
+        }
+        case RegexGroup(_, term) =>
+          isEmptyRepetition(term) 
+        case RegexSequence(parts) =>
+          parts.forall(isEmptyRepetition)
+        // cuDF does not support repetitions adjacent to a choice (eg. "a*|a"), but if
+        // we did, we would need to add a `case RegexChoice()` here
+        case _ => false
+      }
+    }
+
+    // For empty strings and a regex containing only a zero-match reptition, 
+    // the behaviour on Spark is different. 
+    // see https://github.com/NVIDIA/spark-rapids/issues/5456
+    if (RegExpShim.reproduceEmptyStringBug() &&
+      isEmptyRepetition(new RegexParser(javaRegexpPattern).parse())) {
```

Comment:
nit: indent is off per the style guide we follow (see CONTRIBUTING.md for links to style guides)

```suggestion
        isEmptyRepetition(new RegexParser(javaRegexpPattern).parse())) {
```